source/common/logger/ur_sinks.h

```diff
@@ -0,0 +1,77 @@
+// Copyright (C) 2023 Intel Corporation
+// SPDX-License-Identifier: MIT
+
+#include <iostream>
+
+namespace sink {
+
+class Sink {
+  public:
+    template <typename... Args>
+    void log(const char *level, const char *fmt, Args &&...args) {
+        *ostream << "[" << level << "]:";
+        format(fmt, std::forward<Args &&>(args)...);
+        *ostream << "\n";
+        ostream->flush();
+    }
+
+  protected:
+    std::ostream *ostream;
+
+    Sink() = default;
+    virtual ~Sink() = 0;
+
+  private:
+    template <typename Arg> void format(Arg &&arg) { *ostream << arg; }
+
+    template <typename Arg, typename... Args>
+    void format(const char *fmt, Arg &&arg, Args &&...args) {
+        while (*fmt != '{' && *fmt != '\0') {
+            *ostream << *fmt++;
+        }
+
+        if (*fmt == '{') {
+            if (*(++fmt) != '}') {
+                throw std::runtime_error("Only empty braces are allowed!");
+            }
+            *ostream << arg;
+        }
+
+        format(++fmt, std::forward<Args &&>(args)...);
+    }
+};
+Sink::~Sink() {}
+
+class StdoutSink : public Sink {
+  public:
+    StdoutSink() { this->ostream = &std::cout; }
+
+    ~StdoutSink() = default;
+};
+
+class StderrSink : public Sink {
+  public:
+    StderrSink() { this->ostream = &std::cerr; }
+
+    ~StderrSink() = default;
+};
+
+class FileSink : public Sink {
```

Comment:
Can we have a generic StreamSink as well? This would allow as to use it instead with stringstream and format messages for exceptions, etc.